plugins/path_managers/ndiffports.c

```diff
@@ -0,0 +1,647 @@
+        struct ndiffports_subflow_info *sub;
+
+        (void) backup;
+
+        entry = l_hashmap_lookup(ndiffports_connections, L_UINT_TO_PTR(token));
+
+        if (!entry) {
+                l_warn("Closed subflow for unmanaged token %08x", token);
+                return;
+        }
+
+        if (!validate_addrs(entry, laddr, raddr)) {
+                l_warn("Address/token mismatch for token %08x", token);
+                return;
+        }
+
+        sub = find_subflow(entry, laddr);
+
+        if (sub) {
+                if (seconds_elapsed(sub->timestamp) < NDIFFPORTS_REJECT_TIME) {
+                        entry->peer_rejected_consecutive++;
+                }
+
+                sub->in_use = false;
+                sub->timestamp = l_time_now();
+
+                if (entry->active_subflows > 0) {
+                        entry->active_subflows--;
+                } else {
+                        l_error("Underflow when adjusting subflow count");
+                }
+        } else {
+                l_warn("Untracked subflow was closed");
+        }
+
+        if (!entry->server_side &&
+            entry->active_subflows < entry->max_subflows) {
+                add_new_subflow(pm, token, entry);
+        }
+}
+
+/**
+ * Plugin ops to register with the mptcpd core
+ */
+static struct mptcpd_plugin_ops const pm_ops = {
+        .connection_established = ndiffports_connection_established,
+        .connection_closed      = ndiffports_connection_closed,
+        .new_subflow            = ndiffports_new_subflow,
+        .subflow_closed         = ndiffports_subflow_closed,
+};
+
+/**
+ * @brief Ndiffports plugin init hook
+ *
+ * Set up the hashmap for tracking connections.
+ *
+ * @param[in] pm  Path manager object (not used)
+ */
+static int ndiffports_init(struct mptcpd_pm *pm)
+{
+        (void) pm;
+
+        static char const name[] = "ndiffports";
+
+        if (!mptcpd_plugin_register_ops(name, &pm_ops)) {
+                l_error("Failed to initialize "
+                        "n-different-ports "
+                        "path manager plugin.");
+
+                return -1;
+        }
+
+        ndiffports_connections = l_hashmap_new();
+
+        l_info("MPTCP n-different-ports "
+               "path manager initialized.");
+
+        return 0;
+}
+
+/**
+ * @brief Helper for cleaning up connection info
+ *
+ * Delete one connection entry.
+ *
+ * @param[in] pm  Path manager object (not used)
+ */
+static void ndiffports_connection_destroy(void *value)
+{
+        struct ndiffports_connection_info *info = value;
+
+        l_free(info);
+}
+
+/**
+ * @brief Ndiffports plugin exit hook
+ *
+ * Delete the hashmap for tracking connections.
+ *
+ * @param[in] pm  Path manager object (not used)
+ */
+static void ndiffports_exit(struct mptcpd_pm *pm)
+{
+        (void) pm;
+
+        l_hashmap_destroy(ndiffports_connections,
+                          ndiffports_connection_destroy);
```

Comment:
`ndiffports_connection_destroy()` only call `l_free()`.  Why not just pass `l_free()` as the _destroy_ function here and remove `ndiffports_connection_destroy()`?

Reply:
I was initially thinking the connection struct would have some pointers to other dynamically allocated members, but it looks like that won't be needed anytime soon. `l_free()` makes more sense, thanks. Always good to delete code :)

Reply:
Right after I deleted the destroy function, I realized I need it to handle the `l_timeout` I'm adding to the connection struct.